### PR TITLE
#3996: preload starter playground blueprint for unaffiliated users on first run

### LINKED ISSUE #3996: Pre-install Playground blueprint/open Playground on install for individual users
Currently, when an individual onboarding user installs the extension, the installer will open the extension options page or the admin console (which will open the typeform survey), depending on the state of their open tabs at the time.

As part of Onboarding Squad goals for this week, we would like to instead open the playground page with the playground blueprint pre-installed.

**Implementation sketch**
- [x] Pre-install Playground blueprint after extension installation
1. Create a separate file for background script called `onboardingBlueprints.ts` or `preinstallOnboardingBlueprints.ts` inside the `background/` directory
2. Create a function in this file called `preinstallPlaygroundBlueprint` that 
a. makes a network request to the `me` endpoint to verify if the user has a preinstalled blueprints flag set
b. If so, makes a network request to get the Playground recipe by (a hard-coded id?)
c. Installs the recipe via the `installRecipe` action on the `extensionsSlice`
d. Call this function in `background.ts`
- [x] Change `openInstallPage` behavior to open Playground url in a separate tab if the user is an individual user (not enterprise or partner)

**Additional back-end changes**
- Create a flag on the back-end separated from `is_onboarded` that tracks whether or not a user has preinstalled blueprints, captured here: https://github.com/pixiebrix/pixiebrix-app/issues/2126
- Nice to have/future work: create an endpoint that will return the blueprints/recipes to preinstall for use in the background script described above

### PR DESCRIPTION
**Pull request by mnholtz**
## What does this PR do?

- Closes #3996

Remaining todos:
- [x] Because I have time, I will be implementing a starterBlueprints endpoint described in https://github.com/pixiebrix/pixiebrix-app/issues/2126, so as to remove the hard-coded Playground blueprint id

## Reviewer Tips

- Mainly take a look at the `starterBlueprints.ts` background script file. 
- This script fetches starter blueprints, if any, from the new `/onboarding/starter-blueprints/` endpoint, installs them, and opens the Playground page if successful. Otherwise, it does nothing. Only "freemium"/individual users will be getting starter blueprints on installation for the time being.
- All other major changes are on the backend which can be found here: https://github.com/pixiebrix/pixiebrix-app/pull/2133

## Discussion

- I've decided to put starter blueprint installation in the background script, so as to not interfere with the needs of onboarding flows in the future 
- For instance, if starter blueprints were installed on the React side of things, one of our React applications would have to be spun up before any other kind of onboarding view. In this case, we'd like to go straight to the Playground page which is not a React application

## Demo

As an example, I've designated the Amazon Context Menu blueprint as a starter blueprint.

https://user-images.githubusercontent.com/36575242/184456346-12b1073c-0a97-4384-91f8-fbcdab806bf1.mov

## Future Work

- No future work at this time

## Checklist

- [x] Add tests **In progress**
- [x] Designate a primary reviewer @twschiller

### CODE REVIEW

File: src/background/background.ts [hunk 1]
```diff
@@ -58,3 +59,4 @@ initFirefoxCompat();
 activateBrowserActionIcon();
 initActiveTabTracking();
 initPartnerTheme();
+initStarterBlueprints();
```
Comment (twschiller) [inline src/background/background.ts:62]: 👍

File: src/background/starterBlueprints.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+  console.log(profile);
+
+  // 2. If not, fetch the Playground blueprint
+  if (!profile.install_starter_blueprints) {
+    // TODO: uncomment me
+    //return;
+  }
+
+  const { data: playground_blueprint } = await client.get<BlueprintResponse>(
```
Comment (twschiller) [inline src/background/starterBlueprints.ts:51]: Use camelCase in Javascript

File: src/background/starterBlueprints.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+const { reducer, actions } = extensionsSlice;
+// TODO: replace me with the actual playground blueprint id
+const PLAYGROUND_BLUEPRINT_NAME = "google/template-search";
+
+async function installPlaygroundBlueprint(): Promise<void> {
+  // 1. Make a call to the `me` endpoint to see if the user has a falsey preinstalledBlueprints flag
+  const client = await maybeGetLinkedApiClient();
+  if (client == null) {
+    console.debug(
+      "Skipping starter blueprint installation because the extension is not linked to the PixieBrix service"
```
Comment (twschiller) [inline src/background/starterBlueprints.ts:36]: 👍

File: src/background/starterBlueprints.ts [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    })
+  );
+
+  console.log("result", result);
+  await saveOptions(result);
+  await forEachTab(queueReactivateTab);
+
+  // 4. If successful, make a call to the preinstallBlueprints flag endpoint to mark the
+  // preinstalledBlueprints flag
+  await client.post("/api/onboarding/starter-blueprints/", {
```
Comment (twschiller) [inline src/background/starterBlueprints.ts:83]: - This should happen after the saveOptions. (The forEachTab can fail)
- It doesn't need to be awaited

File: src/background/background.ts (hunk 1 above)
Comment (twschiller) [inline src/background/background.ts:62]: 👍

File: src/background/starterBlueprints.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+      "Skipping starter blueprint installation because the extension is not linked to the PixieBrix service"
+    );
+    return;
+  }
+
+  const { data: starterBlueprints } = await client.get<RecipeDefinition[]>(
+    "/api/onboarding/starter-blueprints/"
+  );
+
+  if (!starterBlueprints) {
```
Comment (twschiller) [inline src/background/starterBlueprints.ts:54]: Can this ever be true? starterBlueprints will be `[]` if there are no starter blueprints

File: src/background/starterBlueprints.ts [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      starterBlueprint
+    );
+  }
+
+  await saveOptions(extensionsState);
+
+  void client.post("/api/onboarding/starter-blueprints/install/");
+
+  await forEachTab(queueReactivateTab);
+  window.open("https://www.pixiebrix.com/playground");
```
Comment (twschiller) [inline src/background/starterBlueprints.ts:71]: This is running from the background page. Use the browser.tabs.create API

File: src/background/starterBlueprints.ts [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+import { queueReactivateTab } from "@/contentScript/messenger/api";
+import { ExtensionOptionsState } from "@/store/extensionsTypes";
+
+const { reducer, actions } = extensionsSlice;
+
+function installStarterBlueprint(
+  state: ExtensionOptionsState,
+  starterBlueprint: RecipeDefinition
+): ExtensionOptionsState {
+  return reducer(
```
Comment (twschiller) [inline src/background/starterBlueprints.ts:32]: If the client.post call fails (e.g., due to network error), the next time the extension loads the user will have 2 copies of the blueprint

File: src/background/background.ts (hunk 1 above)
Comment (twschiller) [inline src/background/background.ts:62]: 👍

File: src/background/starterBlueprints.ts [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    );
+    return;
+  }
+
+  try {
+    const { data: starterBlueprints } = await client.get<RecipeDefinition[]>(
+      "/api/onboarding/starter-blueprints/"
+    );
+
+    await client.post("/api/onboarding/starter-blueprints/install/");
```
Comment (twschiller) [inline src/background/starterBlueprints.ts:56]: @mnholtz it would be good to add code comment explaining why comes before the save options

Comment (mnholtz) [reply src/background/starterBlueprints.ts:56]: Ok! I'll keep this in mind

File: src/background/starterBlueprints.test.ts [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+  await saveOptions({
+    extensions: [],
+  });
+
+  isLinkedMock.mockClear();
+  openPlaygroundPage.mockClear();
+});
+
+describe("installStarterBlueprints", () => {
+  test("user has starter blueprints", async () => {
```
Comment (twschiller) [inline src/background/starterBlueprints.test.ts:53]: NIT: the test title is ambiguous, could be clarified to be that the there are starter blueprints available, but that the user has not installed them yet

File: src/background/starterBlueprints.test.ts [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+      },
+    ]);
+
+    axiosMock.onPost().reply(204);
+
+    await installStarterBlueprints();
+    const { extensions } = await loadOptions();
+
+    expect(extensions.length).toBe(1);
+    expect(openPlaygroundPage.mock.calls).toHaveLength(0);
```
Comment (twschiller) [inline src/background/starterBlueprints.test.ts:127]: 💯

**Review by twschiller — APPROVED** [review]

**Pull request merged**